nixos/modules/security/tpm2.nix

```diff
@@ -0,0 +1,172 @@
+{ lib, pkgs, config, ... }:
+let
+  cfg = config.security.tpm2;
+
+  # This snippet is taken from tpm2-tss/dist/tpm-udev.rules, but modified to allow custom user/groups
+  # The idea is that the tssUser is allowed to acess the TPM and kernel TPM resource manager, while
+  # the tssGroup is only allowed to access the kernel resource manager
+  # Therefore, if either of the two are null, the respective part isn't generated
+  udevRules = tssUser: tssGroup: ''
+    ${lib.optionalString (tssUser != null) ''KERNEL=="tpm[0-9]*", MODE="0660", OWNER="${tssUser}"''}
+    ${lib.optionalString (tssUser != null || tssGroup != null)
+      ''KERNEL=="tpmrm[0-9]*", MODE="0660"''
+      + lib.optionalString (tssUser != null) '', OWNER="${tssUser}"''
+      + lib.optionalString (tssGroup != null) '', GROUP="${tssGroup}"''
+     }
+  '';
+
+in {
+  options.security.tpm2 = {
+    enable = lib.mkEnableOption "Trusted Platform Module 2 support";
+
+    tssUser = lib.mkOption {
+      description = "Name of the tpm device-owner and service user, set if applyUdevRules is set";
+      type = lib.types.nullOr lib.types.str;
+      default = if cfg.abrmd.enable then "tss" else "root";
+      defaultText = ''"tss" when using the userspace resource manager, "root" otherwise'';
+    };
+
+    tssGroup = lib.mkOption {
+      description = "Group of the tpm kernel resource manager (tpmrm) device-group, set if applyUdevRules is set";
+      type = lib.types.nullOr lib.types.str;
+      default = "tss";
+    };
+
+    applyUdevRules = lib.mkOption {
+      description = ''
+        Whether to make the /dev/tpm[0-9] devices accessible by the tssUser, or
+        the /dev/tpmrm[0-9] by tssGroup respectively
+      '';
+      type = lib.types.bool;
+      default = true;
+    };
+
+    abrmd = lib.mkOption {
+      description = "TPM2 userspace resource manager service";
+      type = lib.types.submodule ({
+        options = {
+          enable = lib.mkEnableOption "Trusted Platform 2 userspace resource manager daemon";
+
+          package = lib.mkOption {
+            description = "tpm2-abrmd package to use";
+            type = lib.types.package;
+            default = pkgs.tpm2-abrmd;
+            defaultText = "pkgs.tpm2-abrmd";
+          };
+        };
+      });
+    };
+
+    pkcs11 = lib.mkOption {
+      description = "TPM2 PKCS#11 module support";
+      type = lib.types.submodule ({
+        options = {
+          enable = lib.mkEnableOption "TPM2 PKCS#11 tool in system path and shared library in /run/libtpm2_pkcs11.so";
+
+          librarySymlink = lib.mkOption {
+            description = "symlink the tpm2-pkcs11 shared library to /run/libtpm2_pkcs11.so";
```

Comment:
I'm concerned about this:
To use it in application's configuration the PKCS#11 shared library should *really* be symlinked somewhere out of the Nix store. Does it make sense to do that as part of the activation script, and is there a better location?
Of course, you can turn the symlinking off.

Reply:
Why is this needed?

Reply:
For instance, when you want to use the PKCS#11 module for SSH, you'd put this in your config:
```
Host *
        PKCS11Provider /run/libtpm2_pkcs11.so
```
It would be very annoying to put an everchanging path from the nix store in there. I'd much rather use this module, which conveniently symlinks the current nixpkgs' version somewhere. Do you know of a better way to do this?

Reply:
How about using `programs.ssh.extraConfig` directly?

Reply:
I haven't tested this, but it may be an option. Regardless, the entire point of a PKCS#11 shared library is to be found and used by other applications - dynamically, not during build time. Often, configuration files require a path to the library. I should be exposed somewhere, preferably with a constant path pointing to the latest version. Is that something we support in NixOS, and how should I do this.

Reply:
Hmm I see. Okay then I think we can put this library into `environment.systemPackages` such that it's exposed at `/run/current-system/sw/lib/libtmp2_pkcs11.so`, and then drop the `librarySymlink` option because there shouldn't be a problem with always having it.

Reply:
Make sure to use `lib.getLib pkgs.tpm2-pkcs11` (see https://github.com/NixOS/nixpkgs/pull/72374#discussion_r372945734)

Reply:
That works flawlessly, thanks for the awesome hint!